### ISSUE TITLE
skip unit tests if they depend on absent 3rd-party library

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3111]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)| @zooko reported|
|-|-|
|Trac ID|trac#3111|
|Type|enhancement|
|Created|2008-03-22 16:39:13Z|
|Branch|https://github.com/twisted/twisted/tree/without-openssl-3111|

If you run twisted unit tests without installing pyOpenSSL and pyCrypto, then you get errors like these:

```
===============================================================================
[ERROR]: twisted.conch.test.test_userauth

Traceback (most recent call last):
  File "/Users/wonwinmcbrootles/playground/twisted/easy_install-1286-4/twisted/trial/runner.py", line 546, in loadPackage
    module = modinfo.load()
  File "/Users/wonwinmcbrootles/playground/twisted/easy_install-1286-4/twisted/python/modules.py", line 380, in load
    return self.pathEntry.pythonPath.moduleLoader(self.name)
  File "/Users/wonwinmcbrootles/playground/twisted/easy_install-1286-4/twisted/python/modules.py", line 618, in moduleLoader
    return self._moduleLoader(modname)
  File "/Users/wonwinmcbrootles/playground/twisted/easy_install-1286-4/twisted/python/reflect.py", line 426, in namedAny
    topLevelPackage = _importAndCheckStack(trialname)
  File "/Users/wonwinmcbrootles/playground/twisted/easy_install-1286-4/twisted/conch/test/test_userauth.py", line 8, in <module>
    from twisted.conch.ssh import userauth
  File "/Users/wonwinmcbrootles/playground/twisted/easy_install-1286-4/twisted/conch/ssh/userauth.py", line 19, in <module>
    import keys, transport, service
  File "/Users/wonwinmcbrootles/playground/twisted/easy_install-1286-4/twisted/conch/ssh/keys.py", line 19, in <module>
    from Crypto.Cipher import DES3
exceptions.ImportError: No module named Crypto.Cipher
```

It would be nicer if you got SkipTests for those.

<details><summary>Searchable metadata</summary>

```
trac-id__3111 3111
type__enhancement enhancement
reporter__zooko zooko
priority__normal normal
milestone__ 
branch__branches_without_openssl_3111 branches/without-openssl-3111
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__test test
time__1206203953000000 1206203953000000
changetime__1221058741000000 1221058741000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_1"></a>|@jml commented|
|-|-|

bzrlib has a really nice way of doing this. They distinguish between "skipping due to OS missing features" and "skipping due to lack of dependencies". This makes it easier to see which ones you can fix.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [24702]) Branching to 'without-openssl-3111'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

Much better, but not completely fixed.  There are two other tickets, [#2153](https://github.com/twisted/twisted/issues/2153) and [#3419](https://github.com/twisted/twisted/issues/3419), for the two remaining OpenSSL dependencies.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

Build result at http://buildbot.twistedmatrix.com/supported?branch=branches/without-openssl-3111

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @exarkun|
|-|-|

Fine, +1.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [24770]) Merge without-openssl-3111

Author: exarkun
Reviewer: therve
Fixes: [#3111](https://github.com/twisted/twisted/issues/3111)

Adjust almost all of the unit tests so that if either PyCrypto
or pyOpenSSL is unavailable, tests for functionality which
relies on them are skipped.  This removes a number of mostly
uninformative failures from platforms such as PyPy (where
these modules are never available atm) and cleans up the
test results on CPython when one of these dependencies is
missing.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
